helper/querybuilder.php

```diff
@@ -0,0 +1,277 @@
+<?php
+/**
+ * Tagging Plugin (helper component)
+ */
+class helper_plugin_tagging_querybuilder extends DokuWiki_Plugin {
+
+    /** @var string */
+    protected $field;
+    /** @var bool */
+    protected $logicalAnd = false;
+    /** @var array */
+    protected $tags = [];
+    /** @var array  */
+    protected $ns = [];
+    /** @var array */
+    protected $notns = [];
+    /** @var string */
+    protected $pid;
+    /** @var string */
+    protected $tagger = '';
+    /** @var string */
+    protected $limit = '';
+    /** @var string */
+    protected $where;
+    /** @var string */
+    protected $orderby;
+    /** @var string */
+    protected $groupby;
+    /** @var string */
+    protected $having = '';
+    /** @var array */
+    protected $values = [];
+
+    /**
+     * Shorthand method: calls the appropriate getter deduced from $this->field
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getQuery()
+    {
+        if (!$this->field) {
+            throw new \RuntimeException('Failed to build a query, no field specified');
+        }
+        return ($this->field === 'pid') ? $this->getPages() : $this->getTags();
+    }
+
+    /**
+     * Processes all parts of the query for fetching tagged pages
+     *
+     * Returns the query builder object
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getPages()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'pid';
+        $this->orderby = "cnt DESC, pid";
+        if ($this->tags && $this->logicalAnd) $this->having = ' HAVING cnt = ' . count($this->tags);
+
+        return $this;
+    }
+
+    /**
+     * Processes all parts of the query for fetching tags
+     *
+     * Returns the query builder object
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getTags()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'CLEANTAG(tag)';
+        $this->orderby = 'CLEANTAG(tag)';
+
+        return $this;
+    }
+
+    /**
+     * Returns the SQL string
+     * @return string
+     */
+    public function getSql()
```

Comment:
what happens when I call this before calling getPages or getTags?

Reply:
fixed, those methods are gone, see previous comment.